README
Copyright (c) 2007, The Kaphan Foundation.

This file contains instructions for running and configuring the Peerworks Classification Engine.

For information on building and installing the engine see the INSTALL file.

Obtaining an item cache
============================================

TODO: These instructions are probably only relevant for Winnow development.  We should generalize it for others.

With the classifier-tools package it is possible to build an item cache using a collector install as a source.
This is done using the classifier-import script, like so:

  classifier-import --url http://collector.wizztag.org/service --out item_cache_dir --since 10.days

This will download all items fetched in the last 10 days and build and item cache out of them.  The downside
is that this can take quite a while since it is involves lots of HTTP requests that add quite a bit of latency
to the process if it is running over the Internet.  A better and faster solution is to obtain a recent copy
of the item cache and use that.  There will be a relatively up-to-date copy available on S3 at the following URL:

  https://s3.amazonaws.com:443/classifier_db/item_cache.tar.gz

You can download and extract that file and you will have a pre-built and tokenized item cache that you can start using straight
away.

Warning: The above item cache is quite large, about 4GB when uncompressed.  If needed I will provide a smaller version.

Running the Classifier
============================================

The classifier can be run using the "classifier" executable. If you performed the "make install"
of the build instructions this is likely in your PATH.

Run "classifier -h" to see a list of supported arguments.

A typical usage will be to start the classifier on a specified port and item cache directory
and a positive cutoff threshold of 0.9. You probably also want a log file to be generated. This will be enough for most development cases.  

This can be done using the following command:

 classifier --db <item_cache_dir> --port 8008 -t 0.9 --log-file classifier.log

You can then tail -f classifier.log to see the log messages the classifier produces. Once a message appears in the log saying
the classifier has completed initialization you can then access it's http server on port 8008.  Initialization can sometimes
take a few minutes depending on the size of the item cache and whether the file system has cached item cache file access.

Once things are up and running Winnow will be able to talk to the classifier using the standard classifier UI elements.  You can
also do a simple test by entering http://localhost:8009/classifier.xml into a web browser to make sure everything is working.

Shutting it down
=============================================

The classifier is shutdown by sending it a SIGTERM signal. This can be done using CTRL-C or kill <pid>.


HISTORY
=== 1.8.3 (4 June 2010)

* Fixed a bunch of memory leaks.

=== 1.8.2 (21 May 2010)

* Record last used at whenever an item is fetched.

=== 1.8.1 (21 May 2010)

* Fixed bugs in tokenization.
* Changed to libmicrohttpd threaded model.

=== 1.8.0 (20 May 2010)

* Fixed problems on Snow Leopard.
* Update to latest libmicrohttpd.
* Added a C based tokenizer and removed support for external tokenizer service.

=== 1.7.1 (5 March 2009)

* Fixed cache pruning bug using hard coded time.

=== 1.7.0 (19 December 2008)

* Removed the storage of feeds.

=== 1.6.0 (6 November 2008)

* Change item cache to store tokens and xml as blobs in separate SQLite databases.

=== 1.5.2 (12 October 2008)

* Fixed slow point in item tokenization.  Was caused by libcurl sending a Expect: 100-continue header to the tokenizer.

=== 1.5.1 (27 September 2008)

* Fixed deadlock bug when purging old jobs.

=== 1.5.0 (11 August 2008)

* Added support for HMAC authentication.

=== 1.4.3 (7 July 2008)

* Added old job purger.
* Update item cache DB user_version to 3

=== 1.4.2 (25 June 2008)

* Fixed deadlock when adding items.
* Fixed long wait on shutdown.

=== 1.4.1 (24 June 2008) 

* Changed writing of atom documents to use frwite so we don't need to deal with char encoding.

=== 1.4.0 (16 June 2008)

* Restructure and optimization of item cache format.

=== 1.3.0 (4 June 2008)
 
* Support for getting the clues for a tag and item.

=== 1.2.4 (24 May 2008)

 * Make sure to tokenize items that already exist but have no tokens.
 * Loosen up item addition restrictions.  They should only require an id.

=== 1.2.3  (23 May 2008)
 
 * Fixed various memory leaks

=== 1.2.2  (22 May 2008)

 * Fixed memory leak when updating tagger.

=== 1.2.1  (21 May 2008)

 * Fixed bug in setting content length for PUTed taggings.
 * Fixed changing rowid when an item is inserted multiple times. 
 * Removed db id from internal representation of item - just use the string id.

=== 1.2.0  (20 May 2008)

 * Added Tagger Cache.
 * Classifier definitions are fetched via HTTP.
 * Classifier results are sent via HTTP.
 * Removed MySQL dependency.
 * Removed configuration file.

=== 1.1.12 (28 April 2008) 

 * Fixed missing NULL termination of data posted via HTTP.
 * Fixed some compiler warnings.

=== 1.1.11 (17 April 2008) 

 * Fixed random background loading when some items aren't in the in-memory cache.
 
=== 1.1.10 (14 April 2008)

 * Fixed build errors under Leopard.
 * Fixed date parsing under Leopard.
 * Allow addition of items without content in the hope that they will be spidered.
 
=== 1.1.9 (9 April 2008)

 * Don't add items with less than min_tokens tokens in response to Atom PP calls.
 * Fixed a couple more memory leaks.
 * Set the time stamp of items added to a live cache, when they are added to the live cache.
 * Added Capistrano deployment of the classifier.

=== 1.1.8 (7 April 2008)

 * Fixed memory leak when fetching non-memory-cached items during classification.
 
=== 1.1.7 (29 March 2008)

 * Fixed segfault when an invalid item is published.
 
=== 1.1.6 (28 March 2008)

 * Include timestamp in logging.
 * Optimized item loading.
 * Prune small items from cache on load.

=== 1.1.5 (25 March 2008)

 * Fixed about response for classifier.
 * Added memory leak test harness.
 * Fixed memory leaks in item and feed addition.
 
=== 1.1.4 (24 March 2008)

 * Fixed default load_item_since time. Forgotten before.
 * Added some logging to item_cache_load.

=== 1.1.3 (24 March 2008)

 * Fixed feed addition to handle feed without titles.
 * Parameterized wait time for item cache updating. This allows you to set the time to wait for more
   items before triggering a new classification job.  Default is 60 seconds.
 * Added load_item_since parameter. This is the number of days worth of items the classifier
   with load on startup. Default is 30.
 * Added index on entries(updated).
 
=== 1.1.2

 * Classifier handles storage and atomization of item tokens itself. (Prevents database deadlocks)
 * Turn off debug and trace messages unless configured with --enable-debug
 * Reports git version instead of svn version.
 * Prevent adding new items from triggering a new classification job for every tag on every item.
   Instead it will now allow items to build up to 200 before new jobs a created.  If there is a
   lull in the addition of new items (currently only 1 second) it will trigger jobs for what it has
   already added.  1 second might not be enough though so this should probably be parameterized.

=== 1.1.1

 * Handle incremental posting of http data.
 
=== 1.1.0

 * Supports item cache architecture.

=== 1.0

 * Initial release


src/main.c
/* Copyright (c) 2007 The Kaphan Foundation
 *
 * Possession of a copy of this file grants no permission or license
 * to use, modify, or create derivate works.
 *
 * Please contact info@peerworks.org for further information.
 */

#include <pthread.h>
#include <config.h>
#include <stdlib.h>
#include <getopt.h>
#include <stdio.h>
#include <signal.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <json/json.h>
#include "logging.h"
#include "classification_engine.h"
#include "httpd.h"
#include "misc.h"
#include "fetch_url.h"
#include "xml_error_functions.h"

#define DEFAULT_LOG_FILE "classifier.log"
#define DEFAULT_PID_FILE "classifier.pid"
#define DEFAULT_DB_FILE "classifier.db"
#define DEFAULT_CACHE_UPDATE_WAIT_TIME 60
#define DEFAULT_LOAD_ITEMS_SINCE 30
#define DEFAULT_MIN_TOKENS 50

#define PID_VAL 512
#define DB_VAL  513
#define CREATE_DB_VAL 514
#define CACHE_UPDATE_WAIT_TIME_VAL 515
#define LOAD_ITEMS_SINCE_VAL 516
#define MIN_TOKENS_VAL 517
#define PERFORMANCE_LOG_FILE_VAL 519
#define TAG_INDEX_VAL 520

#define SHORT_OPTS "hvdo:t:n:p:a:c:"
#define USAGE "Usage: classifier [-dvh] [-o LOGFILE] [--db DATABASE_FILE] [--pid PIDFILE]  [--create-db]\n"

static ItemCacheOptions item_cache_options;
static ItemCache *item_cache;
static Credentials classifier_credentials = {NULL, NULL};
static Credentials item_cache_credentials = {NULL, NULL};
static Credentials classification_credentials = {NULL, NULL};
static TaggerCacheOptions tagger_cache_options = {NULL, &classifier_credentials};
static TaggerCache *tagger_cache;
static ClassificationEngineOptions ce_options = {1, 0.0, NULL, &classifier_credentials};
static ClassificationEngine *engine;
static Httpd *httpd;
static HttpConfig http_config = {8080, NULL, &item_cache_credentials, &classification_credentials};

static void parse_credential(struct json_object * credentials, Credentials * target, const char * role) {
  struct json_object *role_credentials = NULL;
    
  if ((role_credentials = json_object_object_get(credentials, role))) {
    struct json_object * access_id_json = json_object_object_get(role_credentials, "access_id");
    struct json_object * secret_json = json_object_object_get(role_credentials, "secret_key");

    if (access_id_json && secret_json && 
            json_object_is_type(access_id_json, json_type_string) && 
            json_object_is_type(secret_json, json_type_string)) {
      target->access_id = strdup(json_object_get_string(access_id_json));
      target->secret_key = strdup(json_object_get_string(secret_json));
      info("Credentials '%s:XXXX' supplied for role %s", target->access_id, role);
    } else {
      fprintf(stderr, "Bad credentials format for %s. 'access_id' and 'secret_key' properties required\n");
      exit(1);
    }
  } else {
    info("No credentials supplied for role '%s'.", role);
  }
}

static void parse_credentials(const char * credentials_file) {
  struct json_object *credentials_json = json_object_from_file(credentials_file);
  
  if (is_error(credentials_json)) {
    fprintf(stderr, "Error parsing json credentials in %s", credentials_file);
    exit(1);
  } else {
    parse_credential(credentials_json, &item_cache_credentials, "item_cache");
    parse_credential(credentials_json, &classification_credentials, "classification");
    parse_credential(credentials_json, &classifier_credentials, "classifier");
    json_object_put(credentials_json);
  }
}

static void printHelp(void) {
  printf("This is the Peerwork classifier.\n\n");
  printf("Usage: classifier [OPTIONS]\n\n");
  printf(" General Options\n");
  printf("    -d               runs as a daemon\n");
  printf("    -v, --version    prints version information\n");
  printf("    -h, --help       prints this help text\n");
  printf("    -o, --log-file FILE\n");
  printf("                     location of the file to write log messages to\n");
  printf("                     Default: %s\n", DEFAULT_LOG_FILE);
  printf("        --pid FILE   location to write the pid to when running as a daemon\n");
  printf("                     Default: %s\n\n", DEFAULT_PID_FILE);

  printf(" Classification Options:\n");
  printf("    -n, --worker-threads N\n");
  printf("                     number of threads for processing jobs\n");
  printf("                     Default: 1\n");
  printf("    -t, --positive-threshold N\n");
  printf("                     probability threshold for considering a tag to be\n");
  printf("                     applied to and item\n");
  printf("                     Default: 0\n");
  printf("        --performance-log FILE\n");
  printf("                     location of the file in which to write job timings\n\n");
  printf("        --tag-index URL\n");
  printf("                     URL which provides an index of the tags to classify\n\n");

  printf(" Item Cache Options:\n");
  printf("        --db FILE    location of the item cache database file\n");
  printf("        --create     if provide the classifier with create the database at\n");
  printf("                     --db and exit\n");
  printf("        --cache-update-wait-time N\n");
  printf("                     number of seconds to wait after a cache update\n");
  printf("                     before spawning classification jobs\n");
  printf("                     Default: %i seconds\n", DEFAULT_CACHE_UPDATE_WAIT_TIME);
  printf("        --load-items-since N\n");
  printf("                     how many days back to load items from the item cache\n");
  printf("                     Default: %i days\n", DEFAULT_LOAD_ITEMS_SINCE);
  printf("        --min-tokens N\n");
  printf("                     the minimum number of tokens an item requires to be\n");
  printf("                     classified\n\n");

  printf(" HTTP Options:\n");
  printf("    -p, --port N     the port to run the HTTP server on\n");
  printf("    -a, --allowed_ip IP_ADDRESS\n");
  printf("                     An IP address to allow to make HTTP requests\n");
  printf("                     Default: any\n\n");
}

volatile sig_atomic_t termination_in_progress = 0;

void termination_handler(int sig) {
  if (termination_in_progress) {
    raise(sig);
  } else {
    termination_in_progress = 1;
    fprintf(stderr, "Terminating classifier...\n");

    if (httpd) {
      fprintf(stderr, "\tShutting down httpd.\n");
      httpd_stop(httpd);
    }

    if (engine) {
      fprintf(stderr, "\tShutting down engine.\n");
      ce_stop(engine);
      free_classification_engine(engine);
    }

    if (tagger_cache) {
      fprintf(stderr, "\tClosing tagger cache\n");
      free_tagger_cache(tagger_cache);
    }

    if (item_cache) {
      fprintf(stderr, "\tClosing database.\n");
      free_item_cache(item_cache);
    }

    fprintf(stderr, "Complete.\n");
    exit(sig);
  }
}

static void _daemonize(const char * pid_file) {
  int pid, sid;
  pid = fork();

  if (pid < 0) {
    exit(EXIT_FAILURE);
  } else if (pid > 0) {
    if (pid_file) {
      FILE *pidout = fopen(pid_file, "w");
      if (pidout) {
        fprintf(pidout, "%i", pid);
        fclose(pidout);
      } else {
        fprintf(stderr, "Could not open pid file %s: %s", pid_file, strerror(errno));
      }
    }
    // exit the foreground process
    exit(EXIT_SUCCESS);
  }

  umask(0);

  sid = setsid();
  if (sid < 0) {
    fprintf(stderr, "setsid failed: %s", strerror(errno));
    exit(EXIT_FAILURE);
  }

  if (chdir("/") < 0) {
    fprintf(stderr, "chdir(\"/\") failed: %s", strerror(errno));
    exit(EXIT_FAILURE);
  }
}

static int start_classifier(const char * db_file) {
  SET_XML_ERROR_HANDLERS;

  if (CLASSIFIER_OK != item_cache_create(&item_cache, db_file, &item_cache_options)) {
    fprintf(stderr, "Error opening classifier database file at %s: %s\n", db_file, item_cache_errmsg(item_cache));
    free_item_cache(item_cache);
    return EXIT_FAILURE;
  } else {
    item_cache_load(item_cache);
    item_cache_start_cache_updater(item_cache);
    item_cache_start_purger(item_cache, 60 * 60 * 24);

    tagger_cache = create_tagger_cache(item_cache, &tagger_cache_options);
    tagger_cache->tag_retriever = &fetch_url;
    tagger_cache->tag_index_retriever = &fetch_url;

    info("Fetching tag index for the first time...");
    Array *tags = NULL;
    char *errmsg = NULL;
    int rc = fetch_tags(tagger_cache, &tags, &errmsg);
    if (TAG_INDEX_OK == rc) {
      info("Fetched %i tags from %s", tags->size, tagger_cache_options.tag_index_url);
    } else {
      error("Error fetching tag index: %s", errmsg);
      free(errmsg);
    }

    engine = create_classification_engine(item_cache, tagger_cache, &ce_options);
    httpd = httpd_start(&http_config, engine, item_cache, tagger_cache);
    ce_run(engine);
    return EXIT_SUCCESS;
  }
}

int main(int argc, char **argv) {
  int create_database = false;
  int daemonize = false;
  char *log_file = DEFAULT_LOG_FILE;
  char *pid_file = DEFAULT_PID_FILE;
  char *db_file = DEFAULT_DB_FILE;
  char *credentials_file = NULL;
  char real_log_file[MAXPATHLEN];
  char real_db_file[MAXPATHLEN];
  char real_credentials_file[MAXPATHLEN];
  item_cache_options.cache_update_wait_time = DEFAULT_CACHE_UPDATE_WAIT_TIME;
  item_cache_options.load_items_since = DEFAULT_LOAD_ITEMS_SINCE;
  item_cache_options.min_tokens = DEFAULT_MIN_TOKENS;

  int longindex;
  int opt;
  static struct option long_options[] = {
      {"version", no_argument, 0, 'v'},
      {"help", no_argument, 0, 'h'},

      {"log-file", required_argument, 0, 'o'},
      {"pid", required_argument, 0, PID_VAL},
      {"db", required_argument, 0, DB_VAL},
      {"create-db", no_argument, 0, CREATE_DB_VAL},

      {"cache-update-wait-time", required_argument, 0, CACHE_UPDATE_WAIT_TIME_VAL},
      {"load-items-since", required_argument, 0, LOAD_ITEMS_SINCE_VAL},
      {"min-tokens", required_argument, 0, MIN_TOKENS_VAL},

      {"worker-threads", required_argument, 0, 'n'},
      {"positive-threshold", required_argument, 0, 't'},
      {"performance-log", required_argument, 0, PERFORMANCE_LOG_FILE_VAL},

      {"port", required_argument, 0, 'p'},
      {"allowed_ip", required_argument, 0, 'a'},
      
      {"credentials", required_argument, 0, 'c'},

      {"tag-index", required_argument, 0, TAG_INDEX_VAL},

      {0, 0, 0, 0}
  };

  while (-1 != (opt = getopt_long(argc, argv, SHORT_OPTS, long_options, &longindex))) {
    switch (opt) {
      case 'o':
        log_file = optarg;
        break;
      case PID_VAL:
        pid_file = optarg;
        break;
      case DB_VAL:
        db_file = optarg;
        break;
      case CREATE_DB_VAL:
        create_database = true;
        break;
      case 'd':
        daemonize = true;
        break;

      /* Item Cache Options */
      case CACHE_UPDATE_WAIT_TIME_VAL:
        item_cache_options.cache_update_wait_time = strtol(optarg, NULL, 10);
        break;
      case LOAD_ITEMS_SINCE_VAL:
        item_cache_options.load_items_since = strtol(optarg, NULL, 10);
        break;
      case MIN_TOKENS_VAL:
        item_cache_options.min_tokens = strtol(optarg, NULL, 10);
        break;

      /* Classification Engine Options */
      case 'n': /* Number of worker threads */
        ce_options.worker_threads = strtol(optarg, NULL, 10);
        break;
      case 't': /* Positive threshold for classification */
        ce_options.positive_threshold = strtod(optarg, NULL);
        break;
      case PERFORMANCE_LOG_FILE_VAL:
        ce_options.performance_log = optarg;
        break;

      /* HTTP options */
      case 'p':
        http_config.port = strtol(optarg, NULL, 10);
        break;
      case 'a':
        http_config.allowed_ip = optarg;
        break;
        
      case 'c':
        credentials_file = optarg;
        break;

      /* Tagger Cache options */
      case TAG_INDEX_VAL:
        tagger_cache_options.tag_index_url = optarg;
        break;

      /* Common Options */
      case 'h':
        // TODO Add help
        printHelp();
        exit(0);
      case 'v':
        printf("%s\n", PACKAGE_STRING);
        exit(EXIT_SUCCESS);
        break;
      default:
        fprintf(stderr, USAGE);
        exit(EXIT_FAILURE);
      break;
    }
  }

  int rc = EXIT_SUCCESS;

  if (create_database) {
    char error[512];
    if (CLASSIFIER_OK != item_cache_initialize(db_file, error)) {
      fprintf(stderr, "%s\n", error);
      rc = EXIT_FAILURE;
    } else {
      fprintf(stderr, "Database successfully initialized at '%s'\n", db_file);
    }
  } else {
    if (create_file(log_file)) {
      fprintf(stderr, "Could not create %s: %s\n", log_file, strerror(errno));
      exit(EXIT_FAILURE);
    }

    if (NULL == realpath(log_file, real_log_file)) {
      fprintf(stderr, "Could not find %s: %s\n", real_log_file, strerror(errno));
      exit(EXIT_FAILURE);
    }

    if (NULL == realpath(db_file, real_db_file)) {
      fprintf(stderr, "Could not find %s: %s\n", real_db_file, strerror(errno));
      exit(EXIT_FAILURE);
    }

    if (credentials_file && NULL == realpath(credentials_file, real_credentials_file)) {
      fprintf(stderr, "Could not find %s: %s\n", real_credentials_file, strerror(errno));
      exit(EXIT_FAILURE);
    }
    
    if (daemonize) {
      _daemonize(pid_file);
    }

    if (signal(SIGINT, termination_handler) == SIG_IGN)  signal(SIGINT, SIG_IGN);
    if (signal(SIGHUP, termination_handler) == SIG_IGN)  signal(SIGHUP, SIG_IGN);
    if (signal(SIGTERM, termination_handler) == SIG_IGN) signal(SIGTERM, SIG_IGN);

    initialize_logging(real_log_file);
    
    if (credentials_file) {
      parse_credentials(real_credentials_file);
    }
    
    rc = start_classifier(real_db_file);
  }

  return rc;
}
